mentoring/questionnaire.py

```diff
@@ -0,0 +1,252 @@
+        """
+        return question.human_readable_choices
+
+    def get_tips(self):
+        """
+        Returns the tips contained in this block
+        """
+        tips = []
+        for child_id in self.children:
+            if child_isinstance(self, child_id, TipBlock):
+                tips.append(self.runtime.get_block(child_id))
+        return tips
+
+    def get_submission_display(self, submission):
+        """
+        Get the human-readable version of a submission value
+        """
+        for choice in self.custom_choices:
+            if choice.value == submission:
+                return choice.content
+        return submission
+
+    def author_edit_view(self, context):
+        """
+        Add some HTML to the author view that allows authors to add choices and tips.
+        """
+        fragment = super(QuestionnaireAbstractBlock, self).author_edit_view(context)
+        fragment.add_content(loader.render_template('templates/html/questionnaire_add_buttons.html', {}))
+        fragment.add_css_url(self.runtime.local_resource_url(self, 'public/css/questionnaire-edit.css'))
+        return fragment
+
+    def validate_field_data(self, validation, data):
+        """
+        Validate this block's field data. Instead of checking fields like self.name, check the
+        fields set on data, e.g. data.name. This allows the same validation method to be re-used
+        for the studio editor. Any errors found should be added to "validation".
+
+        This method should not return any value or raise any exceptions.
+        All of this XBlock's fields should be found in "data", even if they aren't being changed
+        or aren't even set (i.e. are defaults).
+        """
+        super(QuestionnaireAbstractBlock, self).validate_field_data(validation, data)
+
+        def add_error(msg):
+            validation.add(ValidationMessage(ValidationMessage.ERROR, msg))
+        if not data.name:
+            add_error(self._(u"A unique Question ID is required."))
+        elif ' ' in data.name:
+            add_error(self._(u"Question ID should not contain spaces."))
+
+    def validate(self):
+        """
+        Validates the state of this XBlock.
+        """
+        validation = super(QuestionnaireAbstractBlock, self).validate()
+
+        def add_error(msg):
+            validation.add(ValidationMessage(ValidationMessage.ERROR, msg))
+
+        # Validate the choice values:
+        all_choice_values = self.all_choice_values
+        all_choice_values_set = set(all_choice_values)
+        if len(all_choice_values) != len(all_choice_values_set):
+            add_error(self._(u"Some choice values are not unique."))
+        # Validate the tips:
+        values_with_tips = set()
+        for tip in self.get_tips():
+            values = set(tip.values)
+            for dummy in (values & values_with_tips):
+                add_error(self._(u"Multiple tips configured for the same choice."))
```

Comment:
This is in a for-loop, but the message is a constant.  Won't this possibly add many copies of the message?  Perhaps it should be:
```
if values & values_with_tips:
    add_error(....)
    break
```